pkg/openscap/openscap.go

```diff
@@ -0,0 +1,193 @@
+
+// inputCVEFunc provides an injectable way to get the cve file for testing.
+type inputCVEFunc func(int) (string, error)
+
+// chrootOscapFunc provides an injectable way to chroot and execute oscap for testing.
+type chrootOscapFunc func(...string) ([]byte, error)
+
+// setEnvFunc provides an injectable way to get the cve file for testing.
+type setEnvFunc func() error
+
+type defaultOSCAPScanner struct {
+	// CVEDir is the directory where the CVE file is saved
+	CVEDir string
+	// ResultsDir is the directory to which the arf report will be written
+	ResultsDir string
+
+	// Image is the metadata of the inspected image
+	image *docker.Image
+	// ImageMountPath is the path where the image to be scanned is mounted
+	imageMountPath string
+
+	rhelDist    rhelDistFunc
+	inputCVE    inputCVEFunc
+	chrootOscap chrootOscapFunc
+	setEnv      setEnvFunc
+}
+
+// NewDefaultScanner returns a new OpenSCAP scanner
+func NewDefaultScanner(cveDir string, resultsDir string) Scanner {
+	scanner := &defaultOSCAPScanner{
+		CVEDir:     cveDir,
+		ResultsDir: resultsDir,
+	}
+
+	scanner.rhelDist = scanner.getRHELDist
+	scanner.inputCVE = scanner.getInputCVE
+	scanner.chrootOscap = scanner.oscapChroot
+	scanner.setEnv = scanner.setOscapChrootEnv
+
+	return scanner
+}
+
+func (s *defaultOSCAPScanner) getRHELDist() (int, error) {
+	for _, dist := range RHELDistNumbers {
+		output, err := s.chrootOscap("oval", "eval", "--id",
+			fmt.Sprintf("%s%d", CPE, dist), CPEDict)
+		if err != nil {
+			return 0, err
+		}
+		if strings.Contains(string(output), fmt.Sprintf("%s%d: true", CPE, dist)) {
+			return dist, nil
+		}
+	}
+	return 0, fmt.Errorf("could not find RHEL dist")
+}
+
+func (s *defaultOSCAPScanner) getInputCVE(dist int) (string, error) {
+	cveName := fmt.Sprintf(DistCVENameFmt, dist)
+	cveFileName := path.Join(s.CVEDir, cveName)
+	cveURL, _ := url.Parse(CVEUrl)
+	cveURL.Path = path.Join(cveURL.Path, cveName)
+
+	out, err := os.Create(cveFileName)
+	if err != nil {
+		return "", fmt.Errorf("Could not create file %s: %v\n", cveFileName, err)
+	}
+	defer out.Close()
+
+	resp, err := http.Get(cveURL.String())
+	if err != nil {
+		return "", fmt.Errorf("Could not download file %s: %v\n", cveURL, err)
+	}
+	defer resp.Body.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	return cveFileName, err
+}
+
+func (s *defaultOSCAPScanner) setOscapChrootEnv() error {
+	for k, v := range map[string]string{
+		"OSCAP_PROBE_ROOT":              s.imageMountPath,
+		"OSCAP_PROBE_OS_VERSION":        "4.2.6-301.fc23.x86_64", // FIXME place holder value
+		"OSCAP_PROBE_ARCHITECTURE":      s.image.Config.Labels["Architecture"],
+		"OSCAP_PROBE_OS_NAME":           Linux,
+		"OSCAP_PROBE_PRIMARY_HOST_NAME": fmt.Sprintf("docker-image-%s", s.image.ID[:ImageShortIDLen]),
+	} {
+		err := os.Setenv(k, v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Wrapper function for executing oscap
+func (s *defaultOSCAPScanner) oscapChroot(oscapArgs ...string) ([]byte, error) {
```

Comment:
Maybe the name is throwing me off here but is this supposed to call a `syscall.Chroot`?